// call-func/build/cjs/src/index.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.throttle = exports.register = exports.isRequired = exports.getEventKey = exports.defaultCall = exports["default"] = exports.debounce = exports.cleanAllRegister = void 0;
var _isRequired = _interopRequireDefault(require("./isRequired.js"));
exports.isRequired = _isRequired["default"];
var _throttle = _interopRequireDefault(require("./throttle.js"));
exports.throttle = _throttle["default"];
var _debounce = _interopRequireDefault(require("./debounce.js"));
exports.debounce = _debounce["default"];
var _defaultCall = _interopRequireDefault(require("./defaultCall.js"));
exports.defaultCall = _defaultCall["default"];
var _getEventKey = _interopRequireDefault(require("./getEventKey.js"));
exports.getEventKey = _getEventKey["default"];
var _register = require("./register.js");
exports.register = _register.register;
exports.cleanAllRegister = _register.cleanAllRegister;
var _callfunc = _interopRequireDefault(require("./callfunc.js"));
exports["default"] = _callfunc["default"];
