// reshow-constant/build/es/index.mjs
// @ts-check

/**
 * Basic type
 */
export var FUNCTION = "function";
export var NUMBER = "number";
export var OBJECT = "object";
export var STRING = "string";

/**
 * @typedef {T_UNDEFINED|T_NULL|T_FALSE|0} SAFE_UNDEFINED
 */

/**
 * @type undefined
 */
export var T_UNDEFINED = void 0;
/**
 * @type null
 */
export var T_NULL = null;
export var T_TRUE = true;
export var T_FALSE = false;
export var DEFAULT = "default";
export var SYMBOL = "symbol";
export var SCRIPT = "script";
export var UNDEFINED = "undefined";
export var TYPE_ERROR = "TypeError";
export var KEYS = Object.keys;
export var IS_ARRAY = Array.isArray;
/**
 * @param {object|string?=} o
 * @returns {number}
 */
export var OBJ_SIZE = function OBJ_SIZE(o) {
  return o ? KEYS(o).length : 0;
};
/**
 * @returns {any}
 */
export var NEW_OBJ = function NEW_OBJ() {
  return Object.create(null);
};

/**
 * Object.hasOwn()
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn#problematic_cases_for_hasownproperty
 * https://github.com/react-atomic/reshow/issues/118#issuecomment-1203255559
 *
 * @param {object|undefined|null} obj
 * @param {string} key
 * @returns {boolean}
 */
export var HAS = function HAS(obj, key) {
  return !!(obj && Object.prototype.hasOwnProperty.call(obj, key));
};

// reshow specific
export var REAL_TIME_URL = "--rtime-url--";
export var REAL_TIME_DATA_KEY = "--rtime-data--";
