// reshow-runtime/helpers/createForOfIteratorHelper.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("reshow-runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = _createForOfIteratorHelper;

var _reshowConstant = require("reshow-constant");

var _unsupportedIterableToArray = _interopRequireWildcard(require("./unsupportedIterableToArray.js"));

var _getTypeOf = _interopRequireDefault(require("./getTypeOf.js"));

function _createForOfIteratorHelper(o, allowArrayLike) {
  var it;

  if ((0, _getTypeOf["default"])(Symbol) === _reshowConstant.UNDEFINED || o[Symbol.iterator] == _reshowConstant.T_NULL) {
    if ((0, _reshowConstant.IS_ARRAY)(o) || (it = (0, _unsupportedIterableToArray["default"])(o)) || allowArrayLike && o && (0, _getTypeOf["default"])(o.length) === _reshowConstant.NUMBER) {
      if (it) o = it;
      var i = 0;

      var F = function F() {};

      return {
        s: F,
        n: function n() {
          if (i >= o.length) return {
            done: true
          };
          return {
            done: false,
            value: o[i++]
          };
        },
        e: function e(_e) {
          throw _e;
        },
        f: F
      };
    }

    throw new TypeError(_unsupportedIterableToArray.NonIterableError);
  }

  var normalCompletion = true,
      didErr = false,
      err;
  return {
    s: function s() {
      it = o[Symbol.iterator]();
    },
    n: function n() {
      var step = it.next();
      normalCompletion = step.done;
      return step;
    },
    e: function e(_e2) {
      didErr = true;
      err = _e2;
    },
    f: function f() {
      try {
        if (!normalCompletion && it["return"] != null) it["return"]();
      } finally {
        if (didErr) throw err;
      }
    }
  };
}

module.exports = exports.default;
