// call-func/build/es/src/register.mjs
import _classCallCheck from "reshow-runtime/es/helpers/classCallCheck";
import _createClass from "reshow-runtime/es/helpers/createClass";
import _defineProperty from "reshow-runtime/es/helpers/defineProperty";
import { KEYS } from "reshow-constant";
import callfunc from "./callfunc.mjs";
var count = {
  el: 0,
  opt: 0
};
var allWrapMap = {};
var wrapKey = "data-event-wrap-id";
var initEventWrap = function initEventWrap(el) {
  var wrapId = el.getAttribute ? el.getAttribute(wrapKey) : el[wrapKey];
  if (!wrapId) {
    wrapId = ++count.el;
    if (el.setAttribute) {
      el.setAttribute(wrapKey, wrapId);
    } else {
      el[wrapKey] = wrapId;
    }
  }
  var obj = allWrapMap[wrapId];
  if (!obj) {
    obj = new EventWrap(wrapId, el);
    allWrapMap[wrapId] = obj;
  }
  return obj;
};
var EventWrap = /*#__PURE__*/function () {
  function EventWrap(_id, el) {
    var _this = this;
    _classCallCheck(this, EventWrap);
    _defineProperty(this, "optionMap", {});
    _defineProperty(this, "typeMap", {});
    _defineProperty(this, "addEventListener", function (type, func) {
      var thisOptId = ++count.opt;
      var optionMap = _this.optionMap;
      for (var _len = arguments.length, options = new Array(_len > 2 ? _len - 2 : 0), _key = 2; _key < _len; _key++) {
        options[_key - 2] = arguments[_key];
      }
      optionMap[thisOptId] = [type, func].concat(options);
      if (!_this.typeMap[type]) {
        _this.typeMap[type] = [];
      }
      _this.typeMap[type].push(thisOptId);
      callfunc(_this.el.addEventListener, optionMap[thisOptId], _this.el);
      return thisOptId;
    });
    _defineProperty(this, "removeEventListener", function (typeOrId, func) {
      var optionMap = _this.optionMap;
      var thisOptions;
      var id;
      if (!isNaN(typeOrId) && optionMap[typeOrId]) {
        id = typeOrId;
        thisOptions = optionMap[id];
      } else {
        for (var _len2 = arguments.length, options = new Array(_len2 > 2 ? _len2 - 2 : 0), _key2 = 2; _key2 < _len2; _key2++) {
          options[_key2 - 2] = arguments[_key2];
        }
        thisOptions = [typeOrId, func].concat(options);
      }
      callfunc(_this.el.removeEventListener, thisOptions, _this.el);
      if (id) {
        var type = thisOptions[0];
        var thisTypeMap = _this.typeMap[type] || [];
        _this.typeMap[type] = thisTypeMap.filter(function (item) {
          return item != id;
        });
        delete optionMap[id];
      }
    });
    this.id = _id;
    this.el = el;
  }
  _createClass(EventWrap, [{
    key: "cleanAll",
    value: function cleanAll(type) {
      var _this2 = this;
      var optionMap = this.optionMap;
      if (null != type) {
        if (this.typeMap[type]) {
          this.typeMap[type].forEach(function (key) {
            _this2.removeEventListener(key);
          });
        }
      } else {
        KEYS(optionMap).forEach(function (key) {
          _this2.removeEventListener(key);
        });
      }
    }
  }]);
  return EventWrap;
}();
var register = function register(el) {
  return initEventWrap(el);
};
var cleanAllRegister = function cleanAllRegister(type) {
  KEYS(allWrapMap).forEach(function (key) {
    allWrapMap[key].cleanAll(type);
  });
};
export { register, cleanAllRegister };
